File: bindings/python/examples/input_demo.py
#!/usr/bin/env python3
"""Input-only demo: confirm, text/password/number/textarea, select, fuzzy, date.

Run in a real terminal:

    python examples/input_demo.py
"""

from __future__ import annotations

import sys

import sparcli as sc


def main() -> int:
    if not sc.input_available():
        sc.println("No interactive terminal – run this directly in a shell.",
                   sc.Style.dim())
        return 1

    sc.rule("sparcli – input demo", sc.RuleOpts(type=sc.BorderType.DOUBLE))

    if sc.confirm("Proceed?", sc.ConfirmOpts(default_yes=True)) is None:
        sc.println("cancelled", sc.Style.dim())
        return 0

    name = sc.text_input(
        "Your name",
        sc.TextInputOpts(placeholder="Ada", boxed=True, width=40))
    sc.println(f"  -> {name!r}", sc.Style.dim())

    # A custom Ctrl-R shortcut (callback mode) that beeps a message.
    shortcuts = sc.Shortcuts().on_callback(
        sc.key_ctrl("r"),
        lambda: sc.println("  (reload pressed)", sc.Style.dim()),
        name="reload")
    note = sc.text_input(
        "Note (Ctrl-R reload · Ctrl-G editor)",
        sc.TextInputOpts(external_editor=True, shortcuts=shortcuts))
    sc.println(f"  -> {note!r}", sc.Style.dim())

    pw = sc.password_input("Password", sc.PasswordOpts())
    sc.println(f"  -> {len(pw) if pw else 0} chars", sc.Style.dim())

    qty = sc.number_input("Quantity",
                          sc.NumberOpts(min=0, max=10, step=1, initial=3))
    sc.println(f"  -> {qty}", sc.Style.dim())

    color = sc.Select(sc.SelectOpts(prompt="Favourite color"))
    color.add("red").add("green").add("blue")
    sc.println(f"  -> index {color.run_one()}", sc.Style.dim())

    langs = sc.Select(sc.SelectOpts(prompt="Languages you use", multi=True))
    for lang in ("C", "Python", "Rust", "Go", "Zig"):
        langs.add(lang)
    sc.println(f"  -> {langs.run()}", sc.Style.dim())

    fz = sc.Fuzzy(sc.FuzzyOpts(prompt="Pick a fruit"))
    fruits = ("apple", "apricot", "banana", "cherry", "date", "fig", "grape")
    for fruit in fruits:
        fz.add(fruit)
    sc.println(f"  -> index {fz.run()}", sc.Style.dim())

    d = sc.datepicker(opts=sc.DatePickerOpts(prompt="Pick a date"))
    sc.println(f"  -> {d}", sc.Style.dim())
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: bindings/rust/sparcli/examples/input_demo.rs
//! Interactive walkthrough of the input widgets. Run in a real terminal:
//! `cargo run --example input_demo`.

use sparcli::{
    confirm, datepicker, number_input, number_input_text, password_input,
    text_input, textarea, ConfirmOpts, DatePickerOpts, Fuzzy, FuzzyOpts,
    NumberOpts, PasswordOpts, Select, SelectOpts, Shortcuts, TextInputOpts,
    TextareaOpts,
};

fn main() -> sparcli::Result<()> {
    if !sparcli::input_available() {
        eprintln!("No interactive terminal - run this directly in a terminal.");
        return Ok(());
    }

    if let Some(yes) = confirm(
        "Proceed with the demo?",
        ConfirmOpts::new().default_yes(true),
    )? {
        println!("-> {yes}");
    }

    // Text input with a custom Ctrl-E shortcut + external editor (Ctrl-G).
    let sc = Shortcuts::new().on_return(
        sparcli::key_ctrl('e'),
        1,
        Some("edit elsewhere"),
    );
    if let Some(name) = text_input(
        "Your name (Ctrl-G editor, Ctrl-E action)",
        TextInputOpts::new()
            .placeholder("Ada Lovelace")
            .external_editor(true)
            .shortcuts(&sc),
    )? {
        if sc.fired() == 1 {
            println!("-> Ctrl-E action requested");
        } else {
            println!("-> {name:?}");
        }
    }

    if let Some(pw) = password_input("Password", PasswordOpts::new())? {
        println!("-> {} chars", pw.chars().count());
    }

    if let Some(qty) =
        number_input("Quantity", NumberOpts::new().range(0.0, 100.0).step(5.0))?
    {
        println!("-> {qty}");
    }

    // Exact text (never via f64), comma as decimal separator - feed the
    // result to a decimal type for money amounts. start_empty starts with
    // an empty field instead of "0,00".
    if let Some(amount) = number_input_text(
        "Amount",
        NumberOpts::new()
            .decimals(2)
            .step(0.5)
            .decimal_sep(',')
            .start_empty(true),
    )? {
        println!("-> {amount} (exact)");
    }

    if let Some(notes) = textarea(
        "Notes (Ctrl-D submit, Ctrl-G editor)",
        TextareaOpts::new().boxed(48).external_editor(true),
    )? {
        println!("-> {} bytes", notes.len());
    }

    let mut sel = Select::new(SelectOpts::new().prompt("Pick a fruit"));
    sel.add("Apple").add("Banana").add("Cherry");
    if let Some(i) = sel.run_one()? {
        println!("-> index {i}");
    }

    let mut fz = Fuzzy::new(FuzzyOpts::new().prompt("Find a city"));
    for c in ["Tokyo", "London", "Berlin", "Paris", "Oslo"] {
        fz.add(c);
    }
    if let Some(i) = fz.run()? {
        println!("-> index {i}");
    }

    if let Some(d) =
        datepicker(None, DatePickerOpts::new().prompt("Pick a date"))?
    {
        println!("-> {:04}-{:02}-{:02}", d.year, d.month, d.day);
    }

    Ok(())
}
